src/main/scala/com/evolution/scache/LoadingCache.scala

```diff
@@ -481,6 +482,169 @@ private[scache] object LoadingCache {
         }
       }
 
+      override def modify[A](key: K, f: Option[V] => (A, Directive[F, V])): F[(A, Option[F[Unit]])] = {
+        0.tailRecM { counter =>
+          ref
+            .access
+            .flatMap { case (entryRefs, setMap) =>
+              entryRefs
+                .get(key)
+                .fold {
+                  f(None) match {
+                    // No entry present in the map, and we want to add a new one
+                    case (a, put: Directive.Put[F, V]) =>
+                      Ref[F]
+                        .of[EntryState[F, V]](EntryState.Value(entryOf(put.value, put.release)))
+                        .flatMap { entryRef =>
+                          setMap(entryRefs.updated(key, entryRef)).map {
+                            case true =>
+                              (a, none[F[Unit]])
+                                .asRight[Int]
+                            // Failed adding new entry to the map, retrying
+                            case false =>
+                              (counter + 1)
+                                .asLeft[(A, Option[F[Unit]])]
+                          }
+                        }
+                    // No entry present in the map, and we don't want to have any, so exiting
+                    case (a, Directive.Ignore | Directive.Remove) =>
+                      (a, none[F[Unit]])
+                        .asRight[Int]
+                        .pure[F]
+                  }
+                } { entryRef =>
+                  0.tailRecM { counter1 =>
```

Comment:
are you sure you need separate retries on different levels?

Reply:
We want to retry from the top if we get `Removed` inside `entryRef`, otherwise we retry accessing the same ref since it's faster.